### PR TITLE
Update for LoadableImage disk path

### PR DESCRIPTION
**Pull request by aji13187**
# Description

Change is made to update the local url of the image (rather than the portal url) when the image is actually loaded and saved to disk.
The `BasemapStyleInfo::thumbnailUrlChanged()` is emitted when the local path of the image is available.

## Type of change

- [ ] Bug fix
- [ ] New sample implementation
- [x] Sample viewer enhancement
- [ ] Other enhancement

**Platforms tested on**:

- [ ] Windows
- [ ] Android
- [ ] Linux
- [x] macOS
- [ ] iOS

## Checklist

- [x] Runs and compiles on all active platforms as a standalone sample
- [x] Runs and compiles in the sample viewer(s)
- [x] Branch is up to date with the latest main/v.next
- [x] All merge conflicts have been resolved
- [x] Self-review of changes
- [x] There are no warnings related to changes
- [ ] No unrelated changes have been made to any other code or project files
- [ ] Code is commented with correct formatting (CTRL+i)
- [x] All variable and method names are camel case
- [x] There is no leftover commented code
- [x] Screenshots are correct size and display in description tab (500px by 500px, platform agnostic)
- [ ] If adding a new sample, it is added to the sample viewer
- [ ] Cherry-picked to Main branch (if applicable)

### CODE REVIEW

**Comment by aji13187** [comment]
This wont make any difference in the output of the thumbnail images on the screen.
Not sure this will confuse user or need comments.

File: CppSamples/Maps/CreateDynamicBasemapGallery/BasemapStyleListModel.cpp [hunk 1]
```diff
@@ -60,7 +61,11 @@ void BasemapStyleListModel::insertItemsIntoGallery(const QList<BasemapStyleInfo*
     m_previews.clear();
     for (const BasemapStyleInfo* info : infos)
     {
-        m_previews.append(std::move(info));
+        connect(info, &BasemapStyleInfo::thumbnailUrlChanged, this, [this, info]() {
```
Comment (GuillaumeBelz) [inline CppSamples/Maps/CreateDynamicBasemapGallery/BasemapStyleListModel.cpp:64]: Is `info` lifetime is long enough to be used in connection? Maybe moving the index outside the lambda function is safer?

```cpp
const auto index = m_previews.indexOf(info);
connect(info, &BasemapStyleInfo::thumbnailUrlChanged, this, [this, index]() {
...
```

File: CppSamples/Maps/CreateDynamicBasemapGallery/BasemapStyleListModel.cpp [hunk 2]
```diff
@@ -51,6 +51,7 @@ QVariant BasemapStyleListModel::data(const QModelIndex& index, int role) const
 
 int BasemapStyleListModel::rowCount(const QModelIndex& parent) const
 {
+    Q_UNUSED(parent);
```
Comment (JamesMBallard) [inline CppSamples/Maps/CreateDynamicBasemapGallery/BasemapStyleListModel.cpp:54]:
```suggestion
    Q_UNUSED(parent)
```

File: CppSamples/Maps/CreateDynamicBasemapGallery/BasemapStyleListModel.cpp (hunk 1 above)
Comment (JamesMBallard) [reply CppSamples/Maps/CreateDynamicBasemapGallery/BasemapStyleListModel.cpp:64]: I like @GuillaumeBelz's idea, but I would augment that to add that we should just switch to a traditional for loop with index instead of repeatedly calling `indexOf` which is not a cheap call.

**Comment by JamesMBallard** [comment]
>This wont make any difference in the output of the thumbnail images on the screen.
>Not sure this will confuse user or need comments.

If you are confident this makes no difference in the workflow, we should not do it. We don't want to overcomplicate the sample code if we can avoid it.

**Review by JamesMBallard — COMMENTED** [review]
One comment to add on to what Guillaume already said, but we should consider if this is worthwhile to do if it makes no difference.

File: CppSamples/Maps/CreateDynamicBasemapGallery/BasemapStyleListModel.cpp (hunk 1 above)
Comment (aji13187) [reply CppSamples/Maps/CreateDynamicBasemapGallery/BasemapStyleListModel.cpp]: Used traditional for loop

**Comment by aji13187** [comment]
> > This wont make any difference in the output of the thumbnail images on the screen.
> > Not sure this will confuse user or need comments.
> 
> If you are confident this makes no difference in the workflow, we should not do it. We don't want to overcomplicate the sample code if we can avoid it.

The only difference is now is the sample is using the disk url rather than the network portal image url. This will improve the loading speed the thumbnails if more reloading of thumbnails is required in user apps, but not really improves anything in this sample since we do not reload the page.
It may help the user, so I think we can keep this change.

File: CppSamples/Maps/CreateDynamicBasemapGallery/BasemapStyleListModel.cpp [hunk 3]
```diff
@@ -51,16 +51,21 @@ QVariant BasemapStyleListModel::data(const QModelIndex& index, int role) const
...
 
 void BasemapStyleListModel::insertItemsIntoGallery(const QList<BasemapStyleInfo*> infos)
 {
     beginResetModel();
     m_previews.clear();
-    for (const BasemapStyleInfo* info : infos)
+    for (auto index = 0; index < infos.size(); ++index)
     {
-        m_previews.append(std::move(info));
+        m_previews.insert(index, std::move(infos.at(index)));
```
Comment (JamesMBallard) [inline CppSamples/Maps/CreateDynamicBasemapGallery/BasemapStyleListModel.cpp:64]:
```suggestion
        m_previews.insert(index, infos.at(index));
```
Guillaume probably knows better than I do, but I assume infos.at is already an r-value so no need to std::move it.

File: CppSamples/Maps/CreateDynamicBasemapGallery/BasemapStyleListModel.cpp [hunk 4]
```diff
@@ -51,16 +51,21 @@ QVariant BasemapStyleListModel::data(const QModelIndex& index, int role) const
...
+        QModelIndex modelIndex = createIndex(index, 0);
+        connect(m_previews.at(index), &BasemapStyleInfo::thumbnailUrlChanged, this, [this, modelIndex]() {
+            emit dataChanged(modelIndex, modelIndex);
+        });
```
Comment (JamesMBallard) [inline CppSamples/Maps/CreateDynamicBasemapGallery/BasemapStyleListModel.cpp:68]: Use consistent bracket style. Also is it possible to just capture the index initially rather than the QModelIndex?
```suggestion
        connect(m_previews.at(index), &BasemapStyleInfo::thumbnailUrlChanged, this, [this, index]() 
        {
            const QModelIndex modelIndex = createIndex(index, 0);
            emit dataChanged(modelIndex, modelIndex);
        });
```

This code may not work, proceed with caution.

File: CppSamples/Maps/CreateDynamicBasemapGallery/BasemapStyleListModel.cpp (hunk 3 above)
Comment (aji13187) [reply CppSamples/Maps/CreateDynamicBasemapGallery/BasemapStyleListModel.cpp]: Looks like QList::at() is not rvalue, but still `std::move` is not necessary because of Qt's implicit sharing or copy-on-write mechanism.

File: CppSamples/Maps/CreateDynamicBasemapGallery/BasemapStyleListModel.cpp (hunk 4 above)
Comment (aji13187) [reply CppSamples/Maps/CreateDynamicBasemapGallery/BasemapStyleListModel.cpp]: Tested and working.

File: CppSamples/Maps/CreateDynamicBasemapGallery/BasemapStyleListModel.cpp (hunk 3 above)
Comment (GuillaumeBelz) [reply CppSamples/Maps/CreateDynamicBasemapGallery/BasemapStyleListModel.cpp]: > infos.at is already an r-value

It's tricky. An function call is a lvalue if the return type is a lvalue-reference (including const) and a prvalue otherwise. https://en.cppreference.com/w/cpp/language/value_category

`QList::at` returns a const ref (https://doc.qt.io/qt-6/qlist.html#at) so it's a lvalue. Using `std::move` change nothing. 

But there is no copy, because `QList::at` returns a const ref and `QList::insert` takes a const ref.

TL;DR: `const` can be removed

Comment (aji13187) [reply CppSamples/Maps/CreateDynamicBasemapGallery/BasemapStyleListModel.cpp]: Thanks, removed `const`.

**Review by JamesMBallard — APPROVED** [review]

**Review by GuillaumeBelz — APPROVED** [review]
